posts/benchmarking_design_decisions/index.qmd

```diff
@@ -0,0 +1,228 @@
+---
+title: "Using benchmarking to guide the adoption of dependencies in R packages"
+author:
+  - name: "James Mba Azam"
+    orcid: "https://orcid.org/0000-0001-5782-7330"
+date: last-modified
+categories: [benchmarking, R, tidyverse, cli, R packages, software architecture]
+format:
+  html:
+    toc: true
+execute: 
+  cache: true
+---
+
+```{r, echo = FALSE}
+knitr::opts_chunk$set(
+  cache = TRUE,
+  message = FALSE
+)
+```
+
+R package developers are constantly faced with design decisions about user
+friendliness. Such design decisions often include checking that inputs are
+rightly specified and how to fix or signal issues arising from intermediate
+calculations or processes within a function (condition handling).
+
+These design decisions often result in taking on dependencies. Dependencies are
+external packages that enable a function to work properly. For example, the
+`tidyverse` is a collection of packages that are often used together to
+make data analysis easier. The `tidyverse` is a dependency for many
+packages that are built on top of it.
+
+Dependencies should not be adopted blindly without carefully considering the
+overhead they introduce. Overhead here refers to the additional work that a
+package must do to make the dependency work. In this article, we will
+demonstrate how to use (micro)benchmarking[^1] to guide the decision to take on
+dependencies.
+
+For this demonstration, we will look at condition handling and signalling,
+that is, what to do when problems arise during code execution [^2].
+Condition signalling in R is classified into errors and warnings. Errors refer
+to cases where something fatal happens during code execution so that it halts,
+for example, when a function is run without specifying the required arguments.
+
+```{r}
+try(rnorm())
+```
+
+Warnings are used to handle less fatal issues with code execution that do not
+need to halt the process.
+
+```{r}
+c(1:2) > c(1:3)
+```
+
+Base R ships with functionalities for condition handling and signalling. There is
+`stop()` for signaling errors and `warning()` for throwing warnings. These base R
+functions are, however, are not flexible enough for advanced formatting: colour,
+progress bar, contextual pluralisation, and so forth.
```

Comment:
Nitpicking here but: base R is flexible enough for this and this is precisely how cli manages to build on top of it to add extra features.

What about something like:

```suggestion
`stop()` for signaling errors and `warning()` for throwing warnings. These base R
functions do not include desirable advanced formatting features: colour,
progress bar, contextual pluralisation, and so forth.
```

Reply:
Awesome. Thanks.